### ISSUE TITLE
Parallel I/O: "Attempt to use feature that was not turned on" despite parallel libnetcdf4

### ISSUE THREAD

**Issue by dionhaefner** [Unidata/netcdf4-python#883]
I installed netcdf4-python via

```bash
$ USE_SETUPCFG=0 HDF5_DIR=$HOME/.local NETCDF4_DIR=$HOME/.local pip install --no-binary :all: netcdf4
```

Installation went fine and I can run most of the examples, but parallel I/O fails with

```bash
$ mpirun -n 2 python mpi_example.py
Traceback (most recent call last):
  File "mpi_example.py", line 7, in <module>
    info=MPI.Info())
  File "netCDF4/_netCDF4.pyx", line 2135, in netCDF4._netCDF4.Dataset.__init__
  File "netCDF4/_netCDF4.pyx", line 1752, in netCDF4._netCDF4._ensure_nc_success
OSError: [Errno -128] NetCDF: Attempt to use feature that was not turned on when netCDF was built.: b'parallel_test.nc'
Traceback (most recent call last):
  File "mpi_example.py", line 7, in <module>
    info=MPI.Info())
  File "netCDF4/_netCDF4.pyx", line 2135, in netCDF4._netCDF4.Dataset.__init__
  File "netCDF4/_netCDF4.pyx", line 1752, in netCDF4._netCDF4._ensure_nc_success
OSError: [Errno -128] NetCDF: Attempt to use feature that was not turned on when netCDF was built.: b'parallel_test.nc'
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[54119,1],1]
  Exit code:    1
--------------------------------------------------------------------------
```

I have built libnetcdf and libhdf5 from source with parallel capabilities:

```bash
$ nc-config --all

This netCDF 4.6.2 has been built with the following features:

  --cc            -> mpicc
  --cflags        -> -I/Users/dion/.local/include
  --libs          -> -L/Users/dion/.local/lib -lnetcdf

  --has-c++       -> no
  --cxx           ->

  --has-c++4      -> no
  --cxx4          ->

  --has-fortran   -> yes
  --fc            -> mpif90
  --fflags        -> -I/Users/dion/.local/include
  --flibs         -> -L/Users/dion/.local/lib -lnetcdff -lnetcdf -lnetcdf
  --has-f90       -> no
  --has-f03       -> yes

  --has-dap       -> yes
  --has-dap2      -> yes
  --has-dap4      -> yes
  --has-nc2       -> yes
  --has-nc4       -> yes
  --has-hdf5      -> yes
  --has-hdf4      -> no
  --has-logging   -> yes
  --has-pnetcdf   -> no
  --has-szlib     -> no
  --has-cdf5      -> yes
  --has-parallel4 -> yes
  --has-parallel  -> yes

  --prefix        -> /Users/dion/.local
  --includedir    -> /Users/dion/.local/include
  --libdir        -> /Users/dion/.local/lib
  --version       -> netCDF 4.6.2
```

There are no other versions of libnetcdf present on my system, and it seems the correct library is used:

```bash
$ otool -L ~/.virtualenvs/veros/lib/python3.7/site-packages/netCDF4/_netCDF4.cpython-37m-darwin.so
/Users/dion/.virtualenvs/veros/lib/python3.7/site-packages/netCDF4/_netCDF4.cpython-37m-darwin.so:
	/Users/dion/.local/lib/libhdf5_hl.100.dylib (compatibility version 102.0.0, current version 102.1.0)
	/Users/dion/.local/lib/libhdf5.103.dylib (compatibility version 104.0.0, current version 104.0.0)
	/Users/dion/.local/lib/libnetcdf.13.dylib (compatibility version 15.0.0, current version 15.1.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
```

Anything I am missing here, or is this a bug?

### Platform

OSX Mojave
Python 3.7 in a virtualenv
HDF5 1.10.4
netCDF4 4.6.2

**Comment by jswhit** [comment]
Do the netcdf-c parallel tests run?

**Comment by jswhit** [comment]
I don't think so, issue #820 only occurred when pnetcdf and parallel netcdf4 were active in the C-lib (and like you say, it was fixed anyway).  

Are you 100% sure that you are using your python virtual env install when you run `mpirun -n 2 python mpi_example.py`?

**Comment by dionhaefner** [comment]
Positive:

```bash
$ python -c "import netCDF4; print(netCDF4.__file__)"
/Users/dion/.virtualenvs/veros/lib/python3.7/site-packages/netCDF4/__init__.py
$ mpirun -n 2 python mpi_example.py
Traceback (most recent call last):
  File "mpi_example.py", line 7, in <module>
    info=MPI.Info(), format="NETCDF4")
  File "netCDF4/_netCDF4.pyx", line 2135, in netCDF4._netCDF4.Dataset.__init__
  File "netCDF4/_netCDF4.pyx", line 1752, in netCDF4._netCDF4._ensure_nc_success
OSError: [Errno -128] NetCDF: Attempt to use feature that was not turned on when netCDF was built.: b'parallel_test.nc'
Traceback (most recent call last):
  File "mpi_example.py", line 7, in <module>
    info=MPI.Info(), format="NETCDF4")
  File "netCDF4/_netCDF4.pyx", line 2135, in netCDF4._netCDF4.Dataset.__init__
  File "netCDF4/_netCDF4.pyx", line 1752, in netCDF4._netCDF4._ensure_nc_success
OSError: [Errno -128] NetCDF: Attempt to use feature that was not turned on when netCDF was built.: b'parallel_test.nc'
```

**Comment by jswhit** [comment]
I'm stumped then - I can't think of any reason why `nc_create_par` would throw that error when run from the python interface and not from a c program.

**Comment by dionhaefner** [comment]
I installed HDF4 and re-built netcdf-c, now the error has changed:

```bash
$ mpirun -n 2 python mpi_example.py
Traceback (most recent call last):
  File "mpi_example.py", line 7, in <module>
    info=MPI.Info())
  File "netCDF4/_netCDF4.pyx", line 2135, in netCDF4._netCDF4.Dataset.__init__
  File "netCDF4/_netCDF4.pyx", line 1752, in netCDF4._netCDF4._ensure_nc_success
OSError: [Errno -36] NetCDF: Invalid argument: b'parallel_test.nc'
Traceback (most recent call last):
  File "mpi_example.py", line 7, in <module>
    info=MPI.Info())
  File "netCDF4/_netCDF4.pyx", line 2135, in netCDF4._netCDF4.Dataset.__init__
  File "netCDF4/_netCDF4.pyx", line 1752, in netCDF4._netCDF4._ensure_nc_success
OSError: [Errno -36] NetCDF: Invalid argument: b'parallel_test.nc'
```

Seems to me like it *does* try and fall back to netCDF classic?

**Comment by jswhit** [comment]
Confirmed - I just updated the travis test to use netcdf-c 4.6.2 and it failed with "NetCDF: Attempt to use feature that was not turned on when netCDF was built".  netcdf-c 4.6.1 appears to work fine.  @DennisHeimbigner - was there any change to the MPI parallel stuff in 4.6.2?

**Comment by DennisHeimbigner** [comment]
The RELEASENOTES.md file does not show anything obvious.
In looking at nc_create_par, I see these two cases:

#ifndef USE_PNETCDF
    /* PnetCDF is disabled but user wants to create classic file in parallel */
    if (!(cmode & NC_NETCDF4))
        return NC_ENOTBUILT;
#endif

#ifndef USE_NETCDF4
   /* NETCDF4 is disabled but user wants to create NETCDF4 file in parallel */
    if (cmode & NC_NETCDF4)
        return NC_ENOTBUILT;
#endif
Does either apply here?

**Comment by jswhit** [comment]
Looks like if I change the `cmode` in `nc_create_part` to `NC_MPIIO | NC_NETCDF4` (instead of just `NC_MPIIO`) the test passes.  I had left out the `NC_NETCDF4` since there is a call to `nc_set_default_format` before the `nc_create_par` call.    Apparently with version 4.6.2 this is no longer good enough.

The fix is in pull request #884.

**Comment by DennisHeimbigner** [comment]
When to use default format is still unclear to me because it has to be last resort.
However, I would have thought that it should be used here. But there are a number of
changes to the handling of this in master, so it may work there.

**Comment by wkliao** [comment]
`NC_MPIIO` is deprecated in NetCDF-C 4.6.2. It was aliased to `NC_PNETCDF` in the prior versions.
The default format set in `nc_set_default_format()` will kick in, if the `cmode` argument in `nc_create_par()` does not contain any of the following flags `NC_NETCDF4`, `NC_CLASSIC_MODEL`, `NC_64BIT_OFFSET`, or `NC_64BIT_DATA`.

So, if `nc_set_default_format()`  is never called and `cmode` in `nc_create_par()` does not contain any of the above flags, then the file to be created will be in the classic CDF-1 format. In this case, PnetCDF feature in NetCDF-C is expected to be enabled .

**Comment by dionhaefner** [comment]
Can confirm that it works with netcdf-c 4.6.1, in case someone comes across this looking for a workaround.

**Comment by jswhit** [comment]
Merged pull request #884, so master should now work with 4.6.2.

**Comment by gihankarunarathne** [comment]
I'm new to NetCDF4 and after working continuously nearly one week on this issue, trying to make parallel I/O support for a [docker container](https://github.com/wdias/adapter-grid/blob/master/Dockerfile), suddenly end up here. After downgrade to NetCDF-c 4.6.1 version, it seems to be working fine.

Googling also didn't show me this result, I think because of the issue name was not matching to the error output.
May be creating a new issue with the name of "[OSError: [Errno -128] NetCDF: Attempt to use feature that was not turned on when netCDF was built.: b'parallel_test.nc'](https://github.com/Unidata/netcdf4-python/issues/883)" will help someone to land here.

**Comment by jswhit** [comment]
Will make a new release this week to address this and a couple other problems with netcdf-c 4.6.2
